packages/nouns-contracts/contracts/governance/ExcessETHBurner.sol

```diff
@@ -0,0 +1,323 @@
+
+        return distanceFromBurnWindowStart <= burnWindowSize;
+    }
+
+    /**
+     * @notice Returns the next burn window start id.
+     * For a given `currentNounId`, it returns the next id which can be represented by
+     * `initialBurnNounId_` + N * `nounIdsBetweenBurns_` for any N >= 0.
+     */
+    function nextBurnWindowStart(
+        uint64 currentNounId,
+        uint64 initialBurnNounId_,
+        uint64 nounIdsBetweenBurns_
+    ) public pure returns (uint64) {
+        // this could not happen during a burn. here as convenience.
+        if (currentNounId < initialBurnNounId_) return initialBurnNounId_;
+
+        uint64 distanceFromBurnWindowStart = (currentNounId - initialBurnNounId_) % nounIdsBetweenBurns_;
+        return currentNounId - distanceFromBurnWindowStart + nounIdsBetweenBurns_;
+    }
+
+    /**
+     * @notice Get the amount of excess ETH in the Nouns treasury.
+     * Excess ETH is defined as the difference between the current treasury value denominated in ETH,
+     * and the expected treasury value denominated in ETH, which is defined as mean auction price * adjusted total supply.
+     * It returns zero during the waiting period, and when expected treasury value is greater than current value.
+     * If treasury balance in native ETH is less than the excess ETH calculation, it returns the treasury ETH balance.
+     * @dev This version does not support burning excess other than in native ETH, e.g. stETH, rETH, etc.
+     * @dev Reverts if there is not enough auction history to calculate the mean auction price.
+     */
+    function excessETH() public view returns (uint256) {
+        uint256 expectedTreasuryValue = expectedTreasuryValueInETH();
+        uint256 treasuryValue = treasuryValueInETH();
+
+        if (expectedTreasuryValue >= treasuryValue) return 0;
+
+        return min(treasuryValue - expectedTreasuryValue, owner().balance);
+    }
+
+    /**
+     * @notice Get the expected treasury value denomiated in ETH.
+     * Expected value is defined as mean auction price * adjusted total supply.
+     */
+    function expectedTreasuryValueInETH() public view returns (uint256) {
+        return meanAuctionPrice() * dao.adjustedTotalSupply();
+    }
+
+    /**
+     * @notice Get the current treasury value denomiated in ETH.
+     * In addition to native ETH, it also includes stETH, wETH, and rETH balances.
+     * @dev for rETH, it uses the getEthValue() function to convert rETH to ETH.
+     */
+    function treasuryValueInETH() public view returns (uint256) {
+        address owner_ = owner();
+        return owner_.balance + stETH.balanceOf(owner_) + wETH.balanceOf(owner_) + rETHBalanceInETH(owner_);
+    }
+
+    /**
+     * @notice Get the mean auction price of the last `numberOfPastAuctionsForMeanPrice` auctions.
+     * @dev Reverts if there is not enough auction history to calculate the mean auction price.
+     */
+    function meanAuctionPrice() public view returns (uint256) {
```

Comment:
May be worth adding a mean view function to the auction house itself to remove the need to re-loop over all prices. It also feels like it may be desirable for others who want to build on Nouns.